server2/Arcadia.Assistant/Arcadia.Assistant.Sharepoint/Sharepoint.cs

```diff
@@ -0,0 +1,307 @@
+        protected override async Task RunAsync(CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                //ServiceEventSource.Current.ServiceMessage(this.Context, "Working-{0}", ++iterations);
+
+                // request Sharepoint calendars
+                var externalStorage = this.externalStorageProvider();
+                var departments = await this.GetDepartmentsList(cancellationToken);
+                var sharepointCalendars = departments.Distinct().ToDictionary(x => x,
+                    dId => this.GetSharepointCalendarsByDepartment(dId).ToDictionary(x => x,
+                        async cal => await this.GetAllSharepointItemsForCalendar(externalStorage, cal)));
+
+                foreach (var departmentId in departments)
+                {
+                    var departmentEmployes = await this.employees.FindEmployeesAsync(EmployeesQuery.Create().ForDepartment(departmentId), cancellationToken);
+                    var employeeIds = departmentEmployes.Select(x => x.EmployeeId).ToArray();
+
+                    var employeeVacations = await this.vacations.GetCalendarEventsByEmployeeAsync(employeeIds, cancellationToken);
+                    var employeeVacationValues = employeeVacations.Values.SelectMany(x => x).ToDictionary(x => CspCalendarEventIdParser.GetCalendarEventIdFromCspId(x.VacationId, CalendarEventTypes.Vacation), x => x);
+
+                    var employeeWorkouts = await this.workouts.GetCalendarEventsCollectionAsync(employeeIds, cancellationToken);
+                    var employeeWorkoutsValues = employeeWorkouts.Values.SelectMany(x => x).ToDictionary(x => CspCalendarEventIdParser.GetCalendarEventIdFromCspId(x.ChangeId, CalendarEventTypes.Workout), x => x);
+
+                    var employeeSickLeaves = await this.sickLeaves.GetCalendarEventsCollectionAsync(employeeIds, cancellationToken);
+                    var employeeSickLeavesValues = employeeSickLeaves.Values.SelectMany(x => x).ToDictionary(x => CspCalendarEventIdParser.GetCalendarEventIdFromCspId(x.SickLeaveId, CalendarEventTypes.Sickleave), x => x);
+
+                    foreach (var calendar in this.GetSharepointCalendarsByDepartment(departmentId))
+                    {
+                        var storageItems = await sharepointCalendars[departmentId][calendar];
+
+                        var employeeVacationsStorageItems = storageItems.Where(x => x.Category == CalendarEventTypes.Vacation).ToList();
+                        var employeeWorkoutsStorageItems = storageItems.Where(x => x.Category == CalendarEventTypes.Workout).ToList();
+                        var employeeSickLeavesStorageItems = storageItems.Where(x => x.Category == CalendarEventTypes.Sickleave).ToList();
+
+                        #region vacation synchronization
```

Comment:
There is a thing called 'Methods' in C#. You should use it instead of #regions ;)

Reply:
Yeap, you'r right